### PR TITLE
Feature/events

### PR DESCRIPTION
**Pull request by EroAuditore**
## Private events

Event planner which allows users to create events and then manage user signups. Users can create events and assist an event.

- User sign in
- User sign out
- User can create an event but must be signed in
- Events split into previous events and upcoming events
- User signed in can see the events that have created
- User signed can join events created

### CODE REVIEW

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -0,0 +1,103 @@
+class EventsController < ApplicationController
+  before_action :set_event, only: %i[ show edit update destroy ]
+  before_action :authenticate_user!, except: %i[index show]
+  # GET /events or /events.json
+  def index
+    @events = Event.all
+
+  end
```
Comment (enaburekhan) [inline app/controllers/events_controller.rb]: You should seperate the events in the index action like so:
def index
  @past_events = Event.previous_events
 @upcoming_events = Event.upcoming_events
end

File: app/controllers/events_controller.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+  # GET /events/1 or /events/1.json
+  def show
+    @atndanceUsers = Attendance.where("event_id = ?", params[:id])
+  end
```
Comment (enaburekhan) [inline app/controllers/events_controller.rb]: The contents of the show action should include the following:
@created_events = current_user.events
@past_events = current_user.attended_events.previous
@upcoming_events = current_user.attended_events.upcoming

File: app/models/event.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+class Event < ApplicationRecord
+    has_many :attendees
+    has_many :users, through: :attendees
+    belongs_to :user
```
Comment (enaburekhan) [inline app/models/event.rb:4]: Online 4, this should be:
belongs_to :creator, class_name: "User"

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+  has_many :attendees
+  has_many :events, through: :attendees
+  has_many :user_events, :foreign_key => "user_id", :class_name => "Event"
```
Comment (enaburekhan) [inline app/models/user.rb:8]: You may need to refactor line 8 to reflect this below:
has_many :events, foreign_key: :creator_id

File: app/views/layouts/_navbar.html.erb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+                <%= link_to "Log out", destroy_user_session_path, method: :delete, class:"button is-info" %>
+              </p>
+            <% else %>
+            <p class="control">
+              <%= link_to 'Sign In', new_user_session_path, class: "button is-primary" %>
+            </p>
+						<p class="control">
+              <%= link_to 'Sign Up', new_user_registration_path, class: "button is-primary" %>
+            </p>
+            <% end %>
```
Comment (enaburekhan) [inline app/views/layouts/_navbar.html.erb:34]: Move this logic to a helper. Views are used to render information that became available to them from the corresponding controllers and not for generating it.

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes required:wrench::hammer::full_moon:
**Hi team!
Great job so far:clap::clap:.**

**Major highlights**

- The PR and readme is very descriptive
- Good job for adding password authentication.
- The UI is awesome

However, we still need to make a few changes to get the project approved:muscle:.

**To improve**

- [x] The user's show page displays events the user has created and events that the user is attending to separated in past and upcoming ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)).  Currently, it's been displayed in the wrong page. Kindly fix this.

- [x] EventsController

`index` action

- lists past events, i.e. `@past_events = Event.previous_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). As already highlighted in the PR below. Kindly fix this.

- lists ucomping events, i.e. `@upcoming_events = Event.upcoming_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). Also fix this as already highlighted below.

- [x] UsersController

`show` action

- lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). Kindly fix this as highlighted below.

- lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). Also fix this as highlighted.

- lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). Also fix this as highlighted below.

- [x] `Database`
- the events table has column `creator_id` of type `integer`. Ths is currently missing in the event table. Kindly add it
- an index for `creator_id` column is added. Also add this as well
Why indices for foreign keys are necessary? In short: Since the foreign key columns will be used often to retrieve related information, it's important to index them (queries on those columns will be faster with indexes). More details: check this [article](
https://semaphoreci.com/blog/2017/05/09/faster-rails-is-your-database-properly-indexed.html) and this [article](https://sqlperformance.com/2012/11/t-sql-queries/benefits-indexing-foreign-keys).

- [x] Views

- Move the below highlighted logic to a helper. There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it. Check this [article](http://codefol.io/posts/Where-Do-I-Put-My-Code/) and this [article](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers) if you need a refresher on how to use a helper in rails. See [screen-shot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from pathwright.

- [x] Test
- models have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/)).

- authentication have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
- event managment features have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))

- [x] Remove commented code from route.rb. use git history if you must view previous codes.

**Optional**

- [ ] routes only for used actions, i.e. resources :events, only: [:new, :create, :show, :index]

Make the above-required changes and when you are ready, request another review.

Do have questions/comments? kindly drop them :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy hacking:blush:
Eric Enaburekhan.

**Review by NtwaliHeritier — APPROVED** [review]
### **STATUS: APPROVED✔✔✔✔**

This is an impressive work. Great implementation of tasks and required changes.

You can merge. 👏👏👏

Happy coding💻💻

**Pull request closed without merging**